sstables/proceed_generator.hh

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2021 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <seastar/core/coroutine.hh>
+#include "sstables/consumer.hh"
+
+class proceed_generator {
```

Comment:
Please add documentation about the theory of operation of this thing.

Reply:
Ok

Reply:
I'm also not completely satisfied with this name, maybe we can come up with a better one.
I used "proceed" in the name because we can generate  ```data_consumer::proceed``` values with it.
You can say that the class is a generator, because you have function that yields values one by one, though the yielded values are actually only for controlling when to stop or whether to call primitive_consumer::consume(), and the actual mutation fragments are saved in the flat_mutation_reader::impl buffer.
It's kind of hard to create a general generator because you can customize it so much in c++(@avikivity  had an idea of [something similar](https://github.com/scylladb/scylla/pull/8653#issuecomment-849590716), but as far as I know, we didn't end up deciding on anything yet)